x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ReplaceDataStreamBackingIndexStep.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.core.ilm;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.IndexAbstraction;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.index.Index;
+
+import java.util.Locale;
+import java.util.Objects;
+
+/**
+ * This step replaces a data stream backing index with the target index, as part of the data stream's backing indices.
+ * Eg. if data stream `foo-stream` is backed by indices [`foo-stream-000001`, `foo-stream-000002`] and we'd like to replace the first
+ * generation index, `foo-stream-000001`, with `shrink-foo-stream-000001`, after this step the `foo-stream` data stream will contain
+ * the following indices
+ * <p>
+ * [`shrink-foo-stream-000001`, `foo-stream-000002`]
+ * <p>
+ * The `foo-stream-000001` index will continue to exist but will not be part of the data stream anymore.
+ * <p>
+ * As the last generation is the write index of the data stream, replacing the last generation index is not allowed.
+ * <p>
+ * This is useful in scenarios following a restore from snapshot operation where the restored index will take the place of the source
+ * index in the ILM lifecycle or in the case where we shrink an index and the shrunk index will take the place of the original index.
+ */
+public class ReplaceDataStreamBackingIndexStep extends ClusterStateActionStep {
+    public static final String NAME = "replace-datastream-backing-index";
+    private static final Logger logger = LogManager.getLogger(ReplaceDataStreamBackingIndexStep.class);
+
+    private final String targetIndexPrefix;
+
+    public ReplaceDataStreamBackingIndexStep(StepKey key, StepKey nextStepKey, String targetIndexPrefix) {
+        super(key, nextStepKey);
+        this.targetIndexPrefix = targetIndexPrefix;
+    }
+
+    @Override
+    public boolean isRetryable() {
+        return true;
+    }
+
+    public String getTargetIndexPrefix() {
+        return targetIndexPrefix;
+    }
+
+    @Override
+    public ClusterState performAction(Index index, ClusterState clusterState) {
+        String originalIndex = index.getName();
+        final String targetIndexName = targetIndexPrefix + originalIndex;
+        IndexMetadata targetIndexMetadata = clusterState.metadata().index(targetIndexName);
+
+        IndexMetadata originalIndexMetadata = clusterState.metadata().index(index);
```

Comment:
I think just to be safe we need a `null` check for these two here, probably with a better message than an NPE further down the line

Reply:
Ah I see you already have a null check for the target, we should check the original also (and maybe move the existing null check up here so it doesn't accidentally get used and cause an NPE?)

Reply:
for `ClusterStateActionStep`s we don't do this check as the `ExecuteStepsUpdateTask` performs it before calling `performAction`. 

I think it's alright to keep this as it is for consistency reasons or otherwise update all `ClusterStateActionStep`s.

What do you think?

Reply:
Ah, I misread this, I'll add the check to the step. `ExecuteStepsUpdateTask` does indeed perform the check too, but we do indeed test in every step.